presto-main/src/main/java/com/facebook/presto/resourcemanager/DistributedQueryResource.java

```diff
@@ -0,0 +1,151 @@
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Response;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static com.facebook.presto.server.security.RoleType.ADMIN;
+import static com.facebook.presto.server.security.RoleType.USER;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON_TYPE;
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+
+@Path("/v1/query")
+@RolesAllowed({USER, ADMIN})
+public class DistributedQueryResource
+{
+    private final ResourceManagerClusterStateProvider clusterStateProvider;
+    private final HttpClient httpClient;
+
+    @Inject
+    public DistributedQueryResource(ResourceManagerClusterStateProvider clusterStateProvider, @ForQueryInfo HttpClient httpClient)
+    {
+        this.clusterStateProvider = requireNonNull(clusterStateProvider, "nodeStateManager is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+    }
+
+    @GET
+    public List<BasicQueryInfo> getAllQueryInfo(@QueryParam("state") String stateFilter)
+    {
+        QueryState expectedState = stateFilter == null ? null : QueryState.valueOf(stateFilter.toUpperCase(Locale.ENGLISH));
+        if (stateFilter == null) {
+            return clusterStateProvider.getQueryInfos();
+        }
+        else {
+            ImmutableList.Builder<BasicQueryInfo> builder = new ImmutableList.Builder<>();
+            for (BasicQueryInfo queryInfo : clusterStateProvider.getQueryInfos()) {
+                if (queryInfo.getState() == expectedState) {
+                    builder.add(queryInfo);
+                }
+            }
+            return builder.build();
+        }
+    }
+
+    @GET
+    @Path("{queryId}")
+    public Response getQueryInfo(@PathParam("queryId") QueryId queryId)
+    {
+        return proxyResponse(queryId, "GET", "");
+    }
+
+    @DELETE
+    @Path("{queryId}")
+    public Response cancelQuery(@PathParam("queryId") QueryId queryId)
+    {
+        return proxyResponse(queryId, "DELETE", "");
+    }
+
+    @PUT
+    @Path("{queryId}/killed")
+    public Response killQuery(@PathParam("queryId") QueryId queryId, String message)
+    {
+        return proxyResponse(queryId, "PUT", "killed");
+    }
+
+    @PUT
+    @Path("{queryId}/preempted")
+    public Response preemptQuery(@PathParam("queryId") QueryId queryId, String message)
+    {
+        return proxyResponse(queryId, "PUT", "preempted");
+    }
+
+    private Response proxyResponse(QueryId queryId, String httpMethod, String additionalPath)
```

Comment:
Does this make sure the client ip address is being passed to the request? If not, we should pass that to track the request correctly.

Reply:
Good point, will add a TODO for this.

Reply:
Per previous offline discussion:

> 1. user send request "/v1/query" to RM. 
> 2. RM calls proxyResponse, and figure out the coordinator responsible to the query 
> 3. RM just streaming the response from coordinator (via StreamingJsonResponseHandler)

Step 1 is further split into 1a and 1b :
> step 1 is actually 1a and 1b, 1a is user send request "/v1/query" to a coordinator (any coordinator) which redirects to RM if needed, 1b is user redirected to RM